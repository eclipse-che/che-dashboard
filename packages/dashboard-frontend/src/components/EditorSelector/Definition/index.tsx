/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import React from 'react';

import { EditorDefinitionField } from '@/components/EditorSelector/Definition/DefinitionField';
import { EditorImageField } from '@/components/EditorSelector/Definition/ImageField';

export type Props = {
  editorDefinition: string | undefined;
  editorImage: string | undefined;
  onChange: (editorId: string | undefined, editorImage: string | undefined) => void;
};

export type State = {
  editorDefinition: string | undefined;
  editorImage: string | undefined;
};

export class EditorDefinition extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      editorDefinition: props.editorDefinition,
      editorImage: props.editorImage,
    };
  }

  private handleEditorDefinition(editorDefinition: string | undefined) {
    const { editorImage } = this.state;

    this.setState({ editorDefinition });
    this.props.onChange(editorDefinition, editorImage);
  }

  private handleEditorImage(editorImage: string | undefined) {
    const { editorDefinition } = this.state;

    this.setState({ editorImage });
    this.props.onChange(editorDefinition, editorImage);
  }

  public render() {
    return (
      <Form isHorizontal={true} onSubmit={e => e.preventDefault()}>
        <EditorDefinitionField
          onChange={editorDefinition => this.handleEditorDefinition(editorDefinition)}
        />
        <EditorImageField onChange={editorImage => this.handleEditorImage(editorImage)} />
      </Form>
    );
  }
}
