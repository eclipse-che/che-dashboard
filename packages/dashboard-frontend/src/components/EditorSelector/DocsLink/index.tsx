/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Button, Flex, FlexItem } from '@patternfly/react-core';
import React from 'react';

const DOCS_DEFINING_A_COMMON_IDE =
  'https://eclipse.dev/che/docs/stable/end-user-guide/defining-a-common-ide/';

export class DocsLink extends React.PureComponent {
  public render() {
    return (
      <Flex>
        <FlexItem align={{ default: 'alignRight' }}>
          <Button
            component="a"
            href={DOCS_DEFINING_A_COMMON_IDE}
            variant="link"
            isInline
            target="_blank"
          >
            How to specify and use a custom editor
          </Button>
        </FlexItem>
      </Flex>
    );
  }
}
