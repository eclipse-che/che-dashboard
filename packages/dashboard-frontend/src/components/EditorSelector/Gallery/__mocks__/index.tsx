/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/components/EditorSelector/Gallery';

export class EditorGallery extends React.PureComponent<Props> {
  public render() {
    const { defaultEditorId, selectedEditorId, onSelect } = this.props;

    return (
      <div data-testid="editor-gallery-component">
        <div>Editor Gallery</div>
        <div data-testid="default-editor-id">{defaultEditorId}</div>
        <div data-testid="selected-editor-id">{selectedEditorId}</div>
        <button onClick={() => onSelect('che-incubator/che-code/latest')}>Select Editor</button>
      </div>
    );
  }
}
