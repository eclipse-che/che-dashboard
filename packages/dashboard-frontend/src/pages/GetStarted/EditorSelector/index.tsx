/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Gallery,
  Panel,
  PanelHeader,
  PanelMain,
  PanelMainBody,
  Title,
} from '@patternfly/react-core';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { EditorSelectorEntry } from '@/pages/GetStarted/EditorSelector/Entry';
import { che } from '@/services/models';
import { AppState } from '@/store';
import { selectEditors } from '@/store/Plugins/chePlugins/selectors';

export type Props = MappedProps & {
  selectedEditorId: string;
  onSelect: (editorId: string) => void;
};
export type State = {
  sortedEditorsByName: Map<string, che.Plugin[]>;
};

class EditorSelector extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      sortedEditorsByName: new Map<string, che.Plugin[]>(),
    };
  }

  public componentDidMount(): void {
    this.init();
  }

  private init(): void {
    const { editors } = this.props;

    const versionPriority = ['insider', 'next', 'latest'];
    const sortedEditors = editors.sort((a, b) => {
      if (a.name === b.name) {
        const aPriority = versionPriority.indexOf(a.version);
        const bPriority = versionPriority.indexOf(b.version);

        if (aPriority !== -1 || bPriority !== -1) {
          return aPriority - bPriority;
        }
      }

      return a.id.localeCompare(b.id);
    });

    const sortedEditorsByName = new Map<string, che.Plugin[]>();

    sortedEditors.forEach(editor => {
      const name = editor.name;
      if (!sortedEditorsByName.has(name)) {
        sortedEditorsByName.set(name, []);
      }
      sortedEditorsByName.get(name)?.push(editor);
    });

    this.setState({
      sortedEditorsByName,
    });
  }

  private handleEditorSelect(editorId: string): void {
    this.props.onSelect(editorId);
  }

  private buildEditorCards(): React.ReactNode[] {
    const { selectedEditorId } = this.props;
    const { sortedEditorsByName } = this.state;

    return Array.from(sortedEditorsByName.keys()).map(editorName => {
      // editors same name, different version
      const editorsGroup = sortedEditorsByName.get(editorName);

      if (editorsGroup === undefined) {
        return;
      }

      const groupIcon = editorsGroup[0].icon;
      const groupName = editorsGroup[0].displayName || editorsGroup[0].name;

      return (
        <EditorSelectorEntry
          key={editorName}
          editorsGroup={editorsGroup}
          groupIcon={groupIcon}
          groupName={groupName}
          selectedId={selectedEditorId}
          onSelect={editorId => this.handleEditorSelect(editorId)}
        />
      );
    });
  }

  render(): React.ReactElement {
    return (
      <Panel>
        <PanelHeader>
          <Title headingLevel="h3">Select an Editor</Title>
        </PanelHeader>
        <PanelMain>
          <PanelMainBody>
            <Gallery
              hasGutter={true}
              minWidths={{ default: '170px' }}
              maxWidths={{ default: '170px' }}
            >
              {this.buildEditorCards()}
            </Gallery>
          </PanelMainBody>
        </PanelMain>
      </Panel>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  editors: selectEditors(state),
});

const connector = connect(mapStateToProps);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(EditorSelector);
