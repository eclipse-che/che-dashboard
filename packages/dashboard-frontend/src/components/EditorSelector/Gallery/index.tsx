/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Gallery } from '@patternfly/react-core';
import React from 'react';

import { EditorSelectorEntry } from '@/components/EditorSelector/Gallery/Entry';
import { che } from '@/services/models';

export type Props = {
  defaultEditorId: string;
  editors: che.Plugin[];
  selectedEditorId: string | undefined;
  onSelect: (editorId: string) => void;
};
export type State = {
  selectedId: string;
  sortedEditorsByName: Map<string, che.Plugin[]>;
};

export class EditorGallery extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      selectedId: '', // will be set on component mount
      sortedEditorsByName: new Map<string, che.Plugin[]>(),
    };
  }

  public componentDidMount(): void {
    this.init();
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    if (prevProps.selectedEditorId !== this.props.selectedEditorId) {
      this.init();
    }
  }

  private init(): void {
    const { defaultEditorId, editors, selectedEditorId, onSelect } = this.props;

    const sortedEditors = sortEditors(editors);

    const sortedEditorsByName = new Map<string, che.Plugin[]>();

    let defaultEditor: che.Plugin | undefined;
    let selectedEditor: che.Plugin | undefined;
    sortedEditors.forEach(editor => {
      const name = editor.name;
      if (!sortedEditorsByName.has(name)) {
        sortedEditorsByName.set(name, []);
      }
      sortedEditorsByName.get(name)?.push(editor);

      // find the default editor
      if (editor.id === defaultEditorId) {
        defaultEditor = editor;
      }
      // find the selected editor
      if (editor.id === selectedEditorId) {
        selectedEditor = editor;
      }
    });

    let selectedId: string;
    if (selectedEditor !== undefined) {
      selectedId = selectedEditor.id;
    } else {
      if (defaultEditor !== undefined) {
        selectedId = defaultEditor.id;
      } else {
        selectedId = sortedEditors[0].id;
      }
      onSelect(selectedId);
    }

    this.setState({
      selectedId,
      sortedEditorsByName,
    });
  }

  private handleEditorSelect(editorId: string): void {
    this.props.onSelect(editorId);
  }

  private buildEditorCards(): React.ReactNode[] {
    const { selectedId, sortedEditorsByName } = this.state;

    return Array.from(sortedEditorsByName.keys()).map(editorName => {
      // editors same name, different version
      const editorsGroup = sortedEditorsByName.get(editorName);

      /* c8 ignore next 3 */
      if (editorsGroup === undefined) {
        return;
      }

      const groupIcon = editorsGroup[0].icon;
      const groupName = editorsGroup[0].displayName || editorsGroup[0].name;

      return (
        <EditorSelectorEntry
          key={editorName}
          editorsGroup={editorsGroup}
          groupIcon={groupIcon}
          groupName={groupName}
          selectedId={selectedId}
          onSelect={editorId => this.handleEditorSelect(editorId)}
        />
      );
    });
  }

  public render() {
    return (
      <Gallery hasGutter={true} minWidths={{ default: '170px' }} maxWidths={{ default: '170px' }}>
        {this.buildEditorCards()}
      </Gallery>
    );
  }
}

const VERSION_PRIORITY: ReadonlyArray<string> = ['insiders', 'next', 'latest'];
export function sortEditors(editors: che.Plugin[]) {
  const sorted = editors.sort((a, b) => {
    if (a.name === b.name) {
      const aPriority = VERSION_PRIORITY.indexOf(a.version);
      const bPriority = VERSION_PRIORITY.indexOf(b.version);

      if (aPriority !== -1 && bPriority !== -1) {
        return aPriority - bPriority;
      } else if (aPriority !== -1) {
        return -1;
      } else if (bPriority !== -1) {
        return 1;
      }
    }

    return a.id.localeCompare(b.id);
  });

  return sorted;
}
