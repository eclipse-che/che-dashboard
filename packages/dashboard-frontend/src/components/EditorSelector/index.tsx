/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionToggle,
  Button,
  Flex,
  FlexItem,
  Panel,
  PanelHeader,
  PanelMain,
  PanelMainBody,
  Title,
} from '@patternfly/react-core';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { EditorDefinition } from '@/components/EditorSelector/Definition';
import { EditorGallery } from '@/components/EditorSelector/Gallery';
import { AppState } from '@/store';
import { selectEditors } from '@/store/Plugins/chePlugins/selectors';

const DOCS_DEFINING_A_COMMON_IDE =
  'https://eclipse.dev/che/docs/stable/end-user-guide/defining-a-common-ide/';

type AccordionId = 'selector' | 'definition';

export type Props = MappedProps & {
  defaultEditorId: string;
  onSelect: (editorDefinition: string | undefined, editorImage: string | undefined) => void;
};
export type State = {
  definitionEditorValue: string | undefined;
  definitionImageValue: string | undefined;

  selectorEditorValue: string;

  expandedId: AccordionId | undefined;
};

class EditorSelector extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      definitionEditorValue: undefined,
      definitionImageValue: undefined,

      selectorEditorValue: '',

      expandedId: 'selector',
    };
  }

  private handleSelectorValueChange(editorId: string): void {
    this.setState({ selectorEditorValue: editorId });
    this.props.onSelect(editorId, undefined);
  }

  private handleDefinitionValueChange(
    editorDefinition: string | undefined,
    editorImage: string | undefined,
  ): void {
    this.setState({
      definitionEditorValue: editorDefinition,
      definitionImageValue: editorImage,
    });
    this.props.onSelect(editorDefinition, editorImage);
  }

  private handleToggle(expandedId: AccordionId): void {
    const { onSelect } = this.props;

    this.setState({
      expandedId: this.state.expandedId === expandedId ? this.state.expandedId : expandedId,
    });

    const { definitionEditorValue, definitionImageValue, selectorEditorValue } = this.state;

    if (expandedId === 'selector') {
      onSelect(selectorEditorValue, undefined);
    } else {
      onSelect(definitionEditorValue, definitionImageValue);
    }
  }

  render(): React.ReactElement {
    const { defaultEditorId, editors } = this.props;
    const { definitionEditorValue, definitionImageValue, selectorEditorValue, expandedId } =
      this.state;

    return (
      <Panel>
        <PanelHeader>
          <Title headingLevel="h3">Editor Selector</Title>
        </PanelHeader>
        <PanelMain>
          <PanelMainBody>
            <Accordion asDefinitionList={false}>
              <AccordionItem>
                <AccordionToggle
                  onClick={() => {
                    this.handleToggle('selector');
                  }}
                  isExpanded={expandedId === 'selector'}
                  id="accordion-item-selector"
                >
                  Choose an Editor
                </AccordionToggle>

                <AccordionContent
                  isHidden={expandedId !== 'selector'}
                  data-testid="editor-gallery-content"
                >
                  <Panel>
                    <PanelMain>
                      <PanelMainBody>
                        <EditorGallery
                          defaultEditorId={defaultEditorId}
                          editors={editors}
                          selectedEditorId={selectorEditorValue}
                          onSelect={editorId => this.handleSelectorValueChange(editorId)}
                        />
                      </PanelMainBody>
                    </PanelMain>
                  </Panel>
                </AccordionContent>
              </AccordionItem>

              <AccordionItem>
                <AccordionToggle
                  onClick={() => {
                    this.handleToggle('definition');
                  }}
                  isExpanded={expandedId === 'definition'}
                  id="accordion-item-definition"
                >
                  Use an Editor Definition
                </AccordionToggle>

                <AccordionContent
                  isHidden={expandedId !== 'definition'}
                  data-testid="editor-definition-content"
                >
                  <Panel>
                    <PanelMain>
                      <PanelMainBody>
                        <EditorDefinition
                          editorDefinition={definitionEditorValue}
                          editorImage={definitionImageValue}
                          onChange={(editorDefinition, editorImage) =>
                            this.handleDefinitionValueChange(editorDefinition, editorImage)
                          }
                        />
                      </PanelMainBody>
                    </PanelMain>
                  </Panel>
                </AccordionContent>
              </AccordionItem>
            </Accordion>
            <Flex>
              <FlexItem align={{ default: 'alignRight' }}>
                <Button
                  component="a"
                  href={DOCS_DEFINING_A_COMMON_IDE}
                  variant="link"
                  isInline
                  target="_blank"
                >
                  How to specify and use a custom editor
                </Button>
              </FlexItem>
            </Flex>
          </PanelMainBody>
        </PanelMain>
      </Panel>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  editors: selectEditors(state),
});

const connector = connect(mapStateToProps, null, null, {
  // forwardRef is mandatory for using `@react-mock/state` in unit tests
  forwardRef: true,
});

type MappedProps = ConnectedProps<typeof connector>;
export default connector(EditorSelector);
