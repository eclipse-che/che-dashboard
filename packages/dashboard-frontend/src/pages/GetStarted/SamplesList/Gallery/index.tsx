/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Button,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStatePrimary,
  EmptyStateVariant,
  Gallery,
  Title,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { SampleCard } from '@/pages/GetStarted/SamplesList/Gallery/Card';
import { che } from '@/services/models';
import * as DevfileRegistriesStore from '@/store/DevfileRegistries';
import { DevfileRegistryMetadata, EMPTY_WORKSPACE_TAG } from '@/store/DevfileRegistries/selectors';

export type PluginEditor = che.Plugin & {
  isDefault: boolean;
};

export const VISIBLE_TAGS = ['Community', 'Tech-Preview', 'Devfile.io'];

export type Props = MappedProps & {
  metadataFiltered: DevfileRegistryMetadata[];
  onCardClick: (metadata: DevfileRegistryMetadata) => void;
};

export class SamplesListGallery extends React.PureComponent<Props> {
  private handleCardClick(metadata: DevfileRegistryMetadata): void {
    this.props.onCardClick(metadata);
  }

  private prepareMetadata(): DevfileRegistryMetadata[] {
    function sortByVisibleTag(a: DevfileRegistryMetadata, b: DevfileRegistryMetadata): -1 | 0 | 1 {
      const getVisibleTag = (metadata: DevfileRegistryMetadata) =>
        metadata.tags.filter(tag => VISIBLE_TAGS.includes(tag))[0];
      const tagA = getVisibleTag(a);
      const tagB = getVisibleTag(b);
      if (tagA === tagB) {
        return 0;
      }
      if (tagA === undefined || tagA < tagB) {
        return -1;
      }
      if (tagB === undefined || tagA > tagB) {
        return 1;
      }
      return 0;
    }

    function sortByEmptyWorkspaceTag(
      a: DevfileRegistryMetadata,
      b: DevfileRegistryMetadata,
    ): -1 | 0 | 1 {
      if (a.tags.includes(EMPTY_WORKSPACE_TAG) > b.tags.includes(EMPTY_WORKSPACE_TAG)) {
        return -1;
      }
      if (a.tags.includes(EMPTY_WORKSPACE_TAG) < b.tags.includes(EMPTY_WORKSPACE_TAG)) {
        return 1;
      }
      return 0;
    }

    function sortByDisplayName(a: DevfileRegistryMetadata, b: DevfileRegistryMetadata): -1 | 0 | 1 {
      if (a.displayName < b.displayName) {
        return -1;
      }
      if (a.displayName > b.displayName) {
        return 1;
      }
      return 0;
    }

    return this.props.metadataFiltered
      .sort(sortByDisplayName)
      .sort(sortByVisibleTag)
      .sort(sortByEmptyWorkspaceTag);
  }

  private buildCardsList(): React.ReactElement[] {
    const metadata = this.prepareMetadata();

    return metadata.map(meta => (
      <SampleCard key={meta.links.v2} metadata={meta} onClick={() => this.handleCardClick(meta)} />
    ));
  }

  render(): React.ReactElement {
    const cards = this.buildCardsList();

    if (cards.length === 0) {
      return (
        <EmptyState variant={EmptyStateVariant.full}>
          <EmptyStateIcon icon={SearchIcon} />
          <Title headingLevel="h1">No results found</Title>
          <EmptyStateBody>
            No results match the filter criteria. Clear filter to show results.
          </EmptyStateBody>
          <EmptyStatePrimary>
            <Button variant="link" onClick={(): void => this.props.clearFilter()}>
              Clear filter
            </Button>
          </EmptyStatePrimary>
        </EmptyState>
      );
    }

    return <Gallery hasGutter={true}>{cards}</Gallery>;
  }
}

const connector = connect(null, {
  ...DevfileRegistriesStore.actionCreators,
});

type MappedProps = ConnectedProps<typeof connector>;
export default connector(SamplesListGallery);
