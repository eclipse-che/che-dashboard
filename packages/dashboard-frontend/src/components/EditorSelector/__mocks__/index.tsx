/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/components/EditorSelector';

export default class EditorSelector extends React.PureComponent<Props> {
  render() {
    const { onSelect } = this.props;
    return (
      <div data-testid="editor-selector">
        Editor Selector
        <button onClick={() => onSelect('some/editor/id', 'custom-editor-image')}>
          Select Editor
        </button>
      </div>
    );
  }
}
