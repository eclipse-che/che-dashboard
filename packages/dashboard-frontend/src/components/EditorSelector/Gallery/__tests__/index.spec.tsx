/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { EditorGallery, sortEditors } from '@/components/EditorSelector/Gallery';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';
import { che } from '@/services/models';

jest.mock('@/components/EditorSelector/Gallery/Entry');

const mockOnSelect = jest.fn();

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const defaultEditorId = 'che-incubator/che-code/insiders';
const selectedEditorId = 'che-incubator/che-code/latest';

describe('EditorGallery', () => {
  let editors: che.Plugin[];

  beforeEach(() => {
    editors = [
      {
        id: 'che-incubator/che-code/insiders',
        description:
          'Microsoft Visual Studio Code - Open Source IDE for Eclipse Che - Insiders build',
        displayName: 'VS Code - Open Source',
        links: {
          devfile: '/v3/plugins/che-incubator/che-code/insiders/devfile.yaml',
        },
        name: 'che-code',
        publisher: 'che-incubator',
        type: 'Che Editor',
        version: 'insiders',
        icon: '/v3/images/vscode.svg',
      },
      {
        id: 'che-incubator/che-code/latest',
        description: 'Microsoft Visual Studio Code - Open Source IDE for Eclipse Che',
        displayName: 'VS Code - Open Source',
        links: {
          devfile: '/v3/plugins/che-incubator/che-code/latest/devfile.yaml',
        },
        name: 'che-code',
        publisher: 'che-incubator',
        type: 'Che Editor',
        version: 'latest',
        icon: '/v3/images/vscode.svg',
      },
      {
        id: 'che-incubator/che-idea-server/latest',
        description: 'JetBrains IntelliJ IDEA Ultimate dev server for Eclipse Che - latest',
        // the displayName is missing - increase the test coverage
        // displayName: 'IntelliJ IDEA Ultimate (desktop)',
        links: {
          devfile: '/v3/plugins/che-incubator/che-idea-server/latest/devfile.yaml',
        },
        name: 'che-idea-server',
        publisher: 'che-incubator',
        type: 'Che Editor',
        version: 'latest',
        icon: '/v3/images/intllij-idea.svg',
      },
      {
        id: 'che-incubator/che-idea-server/next',
        description: 'JetBrains IntelliJ IDEA Ultimate dev server for Eclipse Che - next',
        // the displayName is missing - increase the test coverage
        // displayName: 'IntelliJ IDEA Ultimate (desktop)',
        links: {
          devfile: '/v3/plugins/che-incubator/che-idea-server/next/devfile.yaml',
        },
        name: 'che-idea-server',
        publisher: 'che-incubator',
        type: 'Che Editor',
        version: 'next',
        icon: '/v3/images/intllij-idea.svg',
      },
    ];
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot', () => {
    const snapshot = createSnapshot(editors, defaultEditorId, selectedEditorId);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  describe('sortEditors function', () => {
    test('insiders <=> latest', () => {
      const editorA = editors[0]; // insiders
      const editorB = editors[1]; // latest

      const sortedEditors = sortEditors([editorA, editorB]);

      expect(sortedEditors[0].version).toEqual('insiders');
      expect(sortedEditors[1].version).toEqual('latest');
    });

    test('latest <=> insiders', () => {
      const editorA = editors[1]; // latest
      const editorB = editors[0]; // insiders

      const sortedEditors = sortEditors([editorA, editorB]);

      expect(sortedEditors[0].version).toEqual('insiders');
      expect(sortedEditors[1].version).toEqual('latest');
    });

    test('1.0.0 <=> latest', () => {
      const editorA = editors[0];
      editorA.version = '1.0.0';
      editorA.id = `${editorA.publisher}/${editorA.name}/${editorA.version}`;
      const editorB = editors[1]; // latest

      const sortedEditors = sortEditors([editorA, editorB]);

      expect(sortedEditors[0].version).toEqual('latest');
      expect(sortedEditors[1].version).toEqual('1.0.0');
    });

    test('insiders <=> 1.0.0', () => {
      const editorA = editors[0]; // insiders
      const editorB = editors[1];
      editorB.version = '1.0.0';
      editorA.id = `${editorA.publisher}/${editorA.name}/${editorA.version}`;

      const sortedEditors = sortEditors([editorA, editorB]);

      expect(sortedEditors[0].version).toEqual('insiders');
      expect(sortedEditors[1].version).toEqual('1.0.0');
    });
  });

  describe('select editor', () => {
    test('default and selected IDs are the same', () => {
      renderComponent(editors, defaultEditorId, defaultEditorId);

      expect(mockOnSelect).toHaveBeenCalledTimes(0);

      const nextEditor = editors[3];

      const button = screen.getByRole('button', {
        name: `Select ${nextEditor.name} ${nextEditor.version}`,
      });
      button.click();

      expect(mockOnSelect).toHaveBeenCalledTimes(1);
      expect(mockOnSelect).toHaveBeenNthCalledWith(1, nextEditor.id);
    });

    test('selected ID is different from the default ID', () => {
      renderComponent(editors, defaultEditorId, selectedEditorId);

      expect(mockOnSelect).toHaveBeenCalledTimes(0);

      const nextEditor = editors[3];

      const button = screen.getByRole('button', {
        name: `Select ${nextEditor.name} ${nextEditor.version}`,
      });
      button.click();

      expect(mockOnSelect).toHaveBeenCalledTimes(1);
      expect(mockOnSelect).toHaveBeenNthCalledWith(1, nextEditor.id);
    });

    test('selected ID is undefined', () => {
      renderComponent(editors, defaultEditorId, undefined);

      expect(mockOnSelect).toHaveBeenCalledTimes(1);
      expect(mockOnSelect).toHaveBeenNthCalledWith(1, defaultEditorId);

      const nextEditor = editors[3];

      const button = screen.getByRole('button', {
        name: `Select ${nextEditor.name} ${nextEditor.version}`,
      });
      button.click();

      expect(mockOnSelect).toHaveBeenCalledTimes(2);
      expect(mockOnSelect).toHaveBeenNthCalledWith(2, nextEditor.id);
    });

    test(`default ID doesn't match any editor and selected ID is undefined`, () => {
      renderComponent(editors, 'wrong/editor/id', undefined);

      expect(mockOnSelect).toHaveBeenCalledTimes(1);
      expect(mockOnSelect).toHaveBeenNthCalledWith(1, editors[0].id);

      const nextEditor = editors[3];

      const button = screen.getByRole('button', {
        name: `Select ${nextEditor.name} ${nextEditor.version}`,
      });
      button.click();

      expect(mockOnSelect).toHaveBeenCalledTimes(2);
      expect(mockOnSelect).toHaveBeenNthCalledWith(2, nextEditor.id);
    });

    test('should rerender when the selected editor changes', () => {
      const { reRenderComponent } = renderComponent(editors, defaultEditorId, undefined);

      const selectedEditor = screen.getAllByTestId('selected-editor-id')[0];
      expect(selectedEditor).toHaveTextContent(defaultEditorId);

      reRenderComponent(editors, defaultEditorId, selectedEditorId);

      const nextSelectedEditor = screen.getAllByTestId('selected-editor-id')[0];
      expect(nextSelectedEditor).toHaveTextContent(selectedEditorId);
    });
  });
});

function getComponent(
  editors: che.Plugin[],
  defaultId: string,
  selectedId: string | undefined = undefined,
) {
  return (
    <EditorGallery
      defaultEditorId={defaultId}
      editors={editors}
      selectedEditorId={selectedId}
      onSelect={mockOnSelect}
    />
  );
}
