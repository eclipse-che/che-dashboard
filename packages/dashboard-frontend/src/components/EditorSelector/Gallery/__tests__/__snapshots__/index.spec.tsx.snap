// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorGallery snapshot 1`] = `
<div
  className="pf-l-gallery pf-m-gutter"
  style={
    {
      "--pf-l-gallery--GridTemplateColumns--max": "210px",
      "--pf-l-gallery--GridTemplateColumns--min": "210px",
    }
  }
>
  <div
    data-testid="editor-selector-entry"
  >
    <div
      data-testid="selected-editor-id"
    >
      che-incubator/che-code/latest
    </div>
    <div
      data-testid="editor-group-name"
    >
      VS Code - Open Source
    </div>
    <div
      data-testid="che-code insiders"
    >
      <span>
        che-code
      </span>
      <button
        onClick={[Function]}
      >
        Select 
        che-code
         
        insiders
      </button>
    </div>
    <div
      data-testid="che-code latest"
    >
      <span>
        che-code
      </span>
      <button
        onClick={[Function]}
      >
        Select 
        che-code
         
        latest
      </button>
    </div>
  </div>
  <div
    data-testid="editor-selector-entry"
  >
    <div
      data-testid="selected-editor-id"
    >
      che-incubator/che-code/latest
    </div>
    <div
      data-testid="editor-group-name"
    >
      che-idea-server
    </div>
    <div
      data-testid="che-idea-server next"
    >
      <span>
        che-idea-server
      </span>
      <button
        onClick={[Function]}
      >
        Select 
        che-idea-server
         
        next
      </button>
    </div>
    <div
      data-testid="che-idea-server latest"
    >
      <span>
        che-idea-server
      </span>
      <button
        onClick={[Function]}
      >
        Select 
        che-idea-server
         
        latest
      </button>
    </div>
  </div>
</div>
`;
